File: apps/mobibix-android/app/src/main/java/com/aiyal/mobibix/data/network/LoyaltyApi.kt
package com.aiyal.mobibix.data.network

import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.POST
import retrofit2.http.PUT
import retrofit2.http.Path
import retrofit2.http.Query

data class LoyaltySummary(
    val totalCustomers: Int = 0,
    val totalPointsIssued: Int = 0,
    val totalPointsRedeemed: Int = 0
)

data class LoyaltyHistoryItem(
    val id: String,
    val customerName: String,
    val date: String,
    val type: String, // EARNED, REDEEMED
    val points: Int,
    val description: String?
)

data class LoyaltyTransaction(
    val id: String,
    val customerId: String? = null,
    val customerName: String? = null,
    val type: String = "",
    val points: Int = 0,
    val description: String? = null,
    val createdAt: String? = null
)

data class TransactionsResponse(
    val transactions: List<LoyaltyTransaction> = emptyList()
)

data class ManualAdjustmentRequest(
    val customerId: String,
    val points: Int,
    val reason: String,
    val shopId: String? = null
)

// Keep for backward compat but now maps to manual-adjustment
data class AddPointsRequest(
    val shopId: String,
    val customerId: String,
    val points: Int,
    val description: String?
)

data class RedeemPointsRequest(
    val shopId: String,
    val customerId: String,
    val points: Int,
    val description: String?
)

data class LoyaltyConfig(
    val tenantId: String? = null,
    val isEnabled: Boolean = false,
    val earnAmountPerPoint: Int = 100,
    val pointsPerEarnUnit: Int = 1,
    val pointValueInRupees: Double = 1.0,
    val maxRedeemPercent: Int = 20,
    val allowOnRepairs: Boolean = true,
    val allowOnAccessories: Boolean = true,
    val allowOnServices: Boolean = true,
    val expiryDays: Int? = null,
    val allowManualAdjustment: Boolean = true,
    val minInvoiceForEarn: Int? = null
)

data class UpdateLoyaltyConfigResponse(
    val success: Boolean = true,
    val config: LoyaltyConfig? = null
)

interface LoyaltyApi {
    @GET("api/loyalty/summary")
    suspend fun getLoyaltySummary(@Query("shopId") shopId: String): LoyaltySummary

    // Returns all tenant transactions (optional shopId filter)
    @GET("api/loyalty/transactions")
    suspend fun getTransactions(@Query("shopId") shopId: String? = null): TransactionsResponse

    // Manual adjustment — used for add/subtract points
    @POST("api/loyalty/manual-adjustment")
    suspend fun manualAdjustment(@Body request: ManualAdjustmentRequest): ResponseStatus

    @GET("api/loyalty/config")
    suspend fun getLoyaltyConfig(): LoyaltyConfig

    @PUT("api/loyalty/config")
    suspend fun updateLoyaltyConfig(@Body config: LoyaltyConfig): UpdateLoyaltyConfigResponse
}

data class ResponseStatus(
    val success: Boolean,
    val message: String?
)

File: apps/mobibix-web/src/services/loyalty.api.ts
import { authenticatedFetch, extractData } from "./auth.api";

const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_URL || "http://localhost_REPLACED:3000/api";

export interface LoyaltyConfig {
  tenantId: string;
  isEnabled: boolean;
  earnAmountPerPoint: number; // ₹100 = 1 point
  pointsPerEarnUnit: number;
  pointValueInRupees: number; // 1 point = ₹1
  maxRedeemPercent: number; // Max 50%
  allowOnRepairs: boolean;
  allowOnAccessories: boolean;
  allowOnServices: boolean;
  expiryDays?: number | null;
  allowManualAdjustment: boolean;
  minInvoiceForEarn?: number | null;
  shopId?: string;
}

export interface LoyaltyBalance {
  customerId: string;
  balance: number; // In points
  pointValueInRupees: number;
}

export interface ValidateRedemptionRequest {
  customerId: string;
  points: number;
  invoiceSubTotal: number;
  shopId?: string;
}

export interface ValidateRedemptionResponse {
  success: boolean;
  points?: number;
  discountPaise?: number;
  error?: string;
}

export interface LoyaltyTransaction {
  id: string;
  tenantId: string;
  customerId: string;
  points: number;
  type: "EARN" | "REDEEM" | "EXPIRE" | "MANUAL" | "REVERSAL";
  source?: string;
  invoiceId?: string;
  reversalOf?: string;
  note?: string;
  createdBy?: string;
  createdAt: string | Date;
  customer?: {
    name: string;
  };
}

export interface LoyaltySummary {
  totalPointsIssued: number;
  totalPointsRedeemed: number;
  netPointsBalance: number;
  activeCustomersWithPoints: number;
  shopId?: string;
}

/**
 * Get tenant's loyalty summary stats
 */
export async function getLoyaltySummary(
  startDate?: string,
  endDate?: string,
  shopId?: string,
): Promise<LoyaltySummary | null> {
  try {
    let url = `/loyalty/summary`;
    const params = new URLSearchParams();
    if (startDate) params.append("startDate", startDate);
    if (endDate) params.append("endDate", endDate);
    if (shopId) params.append("shopId", shopId);
    if (params.toString()) url += `?${params.toString()}`;

    const response = await authenticatedFetch(url);
    if (!response.ok) return null;
    const data: any = await extractData(response);
    return data;
  } catch (error) {
    console.error("Error fetching loyalty summary:", error);
    return null;
  }
}

/**
 * Get all loyalty transactions for the tenant
 */
export async function getAllLoyaltyTransactions(shopId?: string): Promise<LoyaltyTransaction[]> {
  try {
    let url = "/loyalty/transactions";
    if (shopId) url += `?shopId=${shopId}`;
    const response = await authenticatedFetch(url);
    if (!response.ok) return [];
    const data: any = await extractData(response);
    return data.transactions || [];
  } catch (error) {
    console.error("Error fetching all loyalty transactions:", error);
    return [];
  }
}

/**
 * Get customer's loyalty balance
 */
export async function getCustomerLoyaltyBalance(
  customerId: string,
  shopId?: string,
): Promise<number> {
  try {
    let url = `/loyalty/balance/${customerId}`;
    if (shopId) url += `?shopId=${shopId}`;
    const response = await authenticatedFetch(url);

    if (!response.ok) {
      console.error("Failed to fetch loyalty balance:", {
        status: response.status,
        customerId,
      });
      return 0; // Return 0 if error
    }

    const data: any = await extractData(response);
    return data.balance || 0;
  } catch (error) {
    console.error("Error fetching loyalty balance:", error);
    return 0;
  }
}

/**
 * Get customer's loyalty transaction history
 */
export async function getCustomerLoyaltyHistory(
  customerId: string,
  shopId?: string,
): Promise<LoyaltyTransaction[]> {
  try {
    let url = `/loyalty/history/${customerId}`;
    if (shopId) url += `?shopId=${shopId}`;
    const response = await authenticatedFetch(url);
    if (!response.ok) return [];
    const data: any = await extractData(response);
    return data.transactions || [];
  } catch (error) {
    console.error("Error fetching loyalty history:", error);
    return [];
  }
}


/**
 * Validate if customer can redeem points
 * Returns max allowed points, discount amount, and validation status
 */
export async function validateLoyaltyRedemption(
  request: ValidateRedemptionRequest,
): Promise<ValidateRedemptionResponse> {
  try {
    const response = await authenticatedFetch(`/loyalty/validate-redemption`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      let errorMessage = "Validation failed";
      try {
        const error = await extractData(response);
        errorMessage = (error as any).error || errorMessage;
      } catch (e) {
        // ignore json parse error
      }
      console.error("Validation error:", {
        status: response.status,
        message: errorMessage,
      });
      return {
        success: false,
        error: errorMessage,
      };
    }

    const data: any = await extractData(response);
    return {
      success: data.success !== false,
      points: data.points,
      discountPaise: data.discountPaise,
      error: data.error,
    };
  } catch (error) {
    console.error("Error validating loyalty redemption:", error);
    return {
      success: false,
      error: "Failed to validate redemption",
    };
  }
}

/**
 * Get tenant's loyalty configuration
 */
export async function getLoyaltyConfig(shopId?: string): Promise<LoyaltyConfig | null> {
  try {
    let url = "/loyalty/config";
    if (shopId) url += `?shopId=${shopId}`;
    const response = await authenticatedFetch(url);

    if (!response.ok) {
      console.error("Failed to fetch loyalty config:", {
        status: response.status,
      });
      return null;
    }

    const data: any = await extractData(response);
    return data;
  } catch (error) {
    console.error("Error fetching loyalty config:", error);
    return null;
  }
}

/**
 * Update tenant's loyalty configuration (admin only)
 */
export async function updateLoyaltyConfig(
  config: Partial<LoyaltyConfig>,
): Promise<LoyaltyConfig | null> {
  try {
    const response = await authenticatedFetch(`/loyalty/config`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(config),
    });

    if (!response.ok) {
      console.error("Failed to update loyalty config:", {
        status: response.status,
      });
      return null;
    }

    const data: any = await extractData(response);
    return data.config;
  } catch (error) {
    console.error("Error updating loyalty config:", error);
    return null;
  }
}

/**
 * Create a manual loyalty adjustment (admin only)
 */
export async function createManualAdjustment(request: {
  customerId: string;
  points: number;
  reason: string;
  shopId?: string;
}): Promise<{ success: boolean; error?: string }> {
  try {
    const response = await authenticatedFetch(`/loyalty/manual-adjustment`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(request),
    });

    if (!response.ok) {
      let errorMessage = "Failed to create adjustment";
      try {
        const error = await extractData(response);
        errorMessage = (error as any).message || errorMessage;
      } catch (e) {
        // ignore json parse error
      }
      return {
        success: false,
        error: errorMessage,
      };
    }

    return { success: true };
  } catch (error) {
    console.error("Error creating manual adjustment:", error);
    return {
      success: false,
      error: "Failed to create adjustment",
    };
  }
}
